src/level/level.cpp

```diff
@@ -238,22 +247,112 @@ namespace level {
+                        logger_.info("enemy firing a bullet");
+                        Position bulletPosition = Position(s->getPosition().riga - 1, s->getPosition().colonna);
+                        weapon::Bullet *bullet = new weapon::Bullet(bulletPosition, enums::Direction::UP, s->getAttack());
+                        this->addBullet(bullet);
+                    }
+                }
+            }
+        }
+        else{
+            s->ShootCoolDown();
+        }
+    }
+
     // BUG: non capisco perché ma ora non funziona più nonostante non abbia cambiato nulla e fa attaccare il nemico se il player
     //      si ritrova sulla colonna o sulla riga di tiro per l'enemy, non su entrambe (che vorrebbe dire che è al massimo a distanza 1)
     void Level::enemiesAttack(WINDOW *win, manager::Level *levelManager) {
         for (unsigned int i = 0; i < enemies_.size(); i++) {
-            if ((levelManager->getPlayer()->getPosition().riga - enemies_[i]->getPosition().riga) <= 1 && (levelManager->getPlayer()->getPosition().riga - enemies_[i]->getPosition().riga) >= -1) {
-                if ((levelManager->getPlayer()->getPosition().colonna - enemies_[i]->getPosition().colonna) <= 1 && (levelManager->getPlayer()->getPosition().colonna - enemies_[i]->getPosition().colonna) >= -1) {
-                    enemies_[i]->attack(levelManager->getPlayer());
-                    if (levelManager->getPlayer()->isDead()) {
-                        // TODO (gio?): implementare una finestra di gameover
+            if(enemies_[i]->getEnemyType() == enums::EnemyType::KAMIKAZE){
+                if ((levelManager->getPlayer()->getPosition().riga - enemies_[i]->getPosition().riga) <= 1 && (levelManager->getPlayer()->getPosition().riga - enemies_[i]->getPosition().riga) >= -1) {
+                    if ((levelManager->getPlayer()->getPosition().colonna - enemies_[i]->getPosition().colonna) <= 1 && (levelManager->getPlayer()->getPosition().colonna - enemies_[i]->getPosition().colonna) >= -1) {
+                        enemies_[i]->attack(levelManager->getPlayer());
+                        if (levelManager->getPlayer()->isDead()) {
+                            // TODO (gio?): implementare una finestra di gameover
+                        }
+                        levelManager->getLogQueue()->addEvent("Il nemico e' esploso");
+                        enemies_[i]->clear(win);
+                        delete enemies_[i];
+                        enemies_.remove(i);
                     }
-                    levelManager->getLogQueue()->addEvent("Il nemico e' esploso");
-                    enemies_[i]->clear(win);
-                    delete enemies_[i];
-                    enemies_.remove(i);
                 }
             }
+            else{
+                if(enemies_[i]->getEnemyType() == enums::EnemyType::SHOOTER){
+                    this->enemyShoot((entities::Shooter *) enemies_[i], levelManager);
+                }
+            }
```

Comment:
Invece di scegliere cosa fargli fare a seconda del tipo ti converrebbe mettere in `Enemy` 
una funzione virtuale pura `attack` che contenga questa logica, così non devi andare a gestirti i singoli modi di attaccare in level. 
La gestione dell'attacco dovrebbe essere interno ad enemy.

Stessa cosa con il player. Infatti ora credo che è abbastanza diffusa sta cosa che puoi metterlo su entity.